docker/docker_image_dest.go

```diff
@@ -99,35 +108,50 @@ func (d *dockerImageDestination) PutBlob(digest string, stream io.Reader) error
 	logrus.Debugf("Uploading %s", uploadURL)
 	res, err = d.c.makeRequest("POST", uploadURL, nil, nil)
 	if err != nil {
-		return err
+		return "", -1, err
 	}
 	defer res.Body.Close()
 	if res.StatusCode != http.StatusAccepted {
 		logrus.Debugf("Error initiating layer upload, response %#v", *res)
-		return fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
+		return "", -1, fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
 	}
 	uploadLocation, err := res.Location()
 	if err != nil {
-		return fmt.Errorf("Error determining upload URL: %s", err.Error())
+		return "", -1, fmt.Errorf("Error determining upload URL: %s", err.Error())
 	}
 
 	// FIXME: DELETE uploadLocation on failure
 
+	h := sha256.New()
+	pr, pw := io.Pipe()
+	bufin := bufio.NewReader(stream)
+	mw := io.MultiWriter(h, pw)
+	go func() {
+		bufin.WriteTo(mw)
+		pw.Close()
+	}()
+	hash := h.Sum(nil)
+	computedDigest := "sha256:" + hex.EncodeToString(hash[:])
+
 	locationQuery := uploadLocation.Query()
-	locationQuery.Set("digest", digest)
+	if digest == "" {
+		locationQuery.Set("digest", computedDigest)
+	} else {
+		locationQuery.Set("digest", digest)
+	}
```

Comment:
If we are computing the digest anyway, we can just always use that one.  (Though a sanity check that the two match wouldn’t hurt… it wouldn’t hurt in any of the `PutBlobs`. This again calls for a helper utility of some kind.)